crates/iota-aws-orchestrator/src/build_cache.rs

```diff
@@ -0,0 +1,401 @@
+        }
+    }
+
+    /// Update instances using build cache
+    pub async fn update_with_build_cache(
+        &self,
+        commit: &str,
+        binaries: &[String],
+        instances_without_metrics: Vec<Instance>,
+        repo_name: String,
+    ) -> TestbedResult<()> {
+        // Detect CPU targets for all instances
+        let cpu_to_instances = self
+            .detect_cpu_targets_for_instances(instances_without_metrics)
+            .await?;
+        if cpu_to_instances.is_empty() {
+            display::action("No instances need binaries");
+            return Ok(());
+        }
+
+        // Take the first available build cache server to resolve the commit
+        let cache_server = self
+            .settings
+            .build_cache
+            .as_ref()
+            .and_then(|cache| cache.servers.values().next())
+            .ok_or_else(|| {
+                BuildCacheError::Cache("No build cache servers configured".to_string())
+            })?;
+
+        let cache_client = BuildCacheClient::with_credentials(
+            cache_server.url.as_str(),
+            cache_server.username.clone(),
+            cache_server.password.clone(),
+        )
+        .map_err(|e| BuildCacheError::Cache(format!("Invalid server URL: {e}")))?;
+
+        let resolved_commit = cache_client
+            .resolve_commit(commit)
+            .await
+            .map_err(|e| BuildCacheError::Cache(format!("Failed to resolve commit: {e}")))?;
+        if commit != resolved_commit {
+            display::action(format!(
+                "Requested commit {commit} resolved to actual commit {resolved_commit} by build cache server",
+            ));
+        }
+
+        // Command needs to run from the repository working directory
+        let release_folder = "./target/release";
+
+        // Process each CPU target group
+        for (cpu_target, instances) in &cpu_to_instances {
+            // Get the build cache config for this CPU target
+            let cache_server = self.settings.build_cache_server_for_target(cpu_target)
+                .ok_or_else(|| BuildCacheError::Cache(format!(
+                    "No build cache server configured for CPU target '{cpu_target}' (needed for {} instances). \
+                     Please add a server configuration that includes '{cpu_target}' in its targets list.",
+                    instances.len()
+                )))?;
+
+            display::action(format!(
+                "Updating builds for commit {resolved_commit} (CPU target: {cpu_target}) using server {}",
+                cache_server.url
+            ));
+
+            let cache_client = BuildCacheClient::with_credentials(
+                cache_server.url.as_str(),
+                cache_server.username.clone(),
+                cache_server.password.clone(),
+            )
+            .map_err(|e| {
+                BuildCacheError::Cache(format!("Invalid server URL for {cpu_target}: {e}"))
+            })?;
```

Comment:
Maybe it's better to pre-build the `cache_client`s, so we don't need to re-create it every iteration.